---
title: "tLOH"
author: "Michelle Webb"
date: "4/27/2021"
output: html_document
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{tLOH}
  %\usepackage[UTF-8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library('tLOH')
library('purrr')
library('dplyr')
```
v0.99.2

tLOH: Assessment of evidence for loss of heterozygosity in spatial 
transcriptomics pre-processed data using Bayes factor calculations.

This tool requires data produced with the 10X Genomics Visium 
Spatial Gene Expression platform and processed following the instructions at
https://github.com/USCDTG/spatialAlleleCountPipeline. The input for tLOH is a
directory of per-cluster .csv(s) with the following columns:

"CHR" - chromosome            
"CHR\_withLabel" - chomosome containing the 'chr' prefix            
"POS" - SNP position        
"POS\_oneBefore" - one position before POS
"rsID" - SNP rsID            
"vcfQUAL" - QUAL field from VCF           
"vcf_refNucleotide" - reference nucleotide from the VCF          
"vcf\_altNucleotide" - alternative nucleotide from the VCF       
"INFO" - VCF Information field          
"vcfAF" - VCF allele fraction            
"genotype" - genotype from the VCF, such as '0/1'        
"A" - coverage of A           
"C" - coverage of C       
"G" - coverage of G         
"T" - coverage of T        
"REF" - reference counts           
"ALT" - alternative counts         

The naming convention of these files in the input directory should be
sample\_cluster[clusterNumber]\_alleleCounts.csv
e.g. Slide1\_cluster1_alleleCounts.csv, Slide1\_cluster2\_alleleCounts.csv

This data is obtained using the tLOH pre-processing python pipeline. The 
purpose of this R package is to perform Bayes calculations on the data from the
output .csvs and plot the results. Examples of how to run this tool are below:


1. tLOHCalc() is the main calculation function. It requires a full path to an input directory and sample name.
```{r marginalCalculations, eval=FALSE}

df <- tLOHCalc('/path/to/sampleDirectory',"sampleName")
```

2. plotLOH() is the plotting function. It requires an input dataframe, 
which will be the output of tLOHCalc, and sample name. Example output df from tLOHCalc is provided in an .rda file in /inst/extdata. The dataframe contains 10436 rows and 20 columns. The column descriptions are as follows -

CHR: chromosome
POS: position
REF: reference allele counts
ALT: alternative allele counts
TOTAL: total counts
p(D|het): probability of data given heterozygous event
p(D|loh): probability of data given loh event
p(het|D): probability of data given heterozygous event divided by the addition 
of p(D|het) and p(D|loh)
p(loh|D): probability of data given loh event divided by the addition of 
p(D|het) and p(D|loh)
bayesFactors: Bayes factor value K
inverseBayes: 1/K
LogBayesFactors: log of Bayes factor K
LogInverseBayes: log of 1/K 
Log10BayesFactors: log 10 of Bayes factor K
Log10InverseBayes: log 10 of 1/K
AF: allele fraction
sample: sample name
Cluster: cluster number
Cluster_AF: cluster + AF for plotting y axis
CHR_F: chromosome factor


The function plotLOH returns a list of two plots, which you can visualize 
individually:

```{r plotting, eval=TRUE}

load('../inst/extdata/df.rda')
listOfPlots <- plotLOH(df,'Example')

listOfPlots[[1]]

listOfPlots[[2]]
```

